### PR TITLE
Class models: `Realm.Object#constructor` for JSI / Hermes

### LINKED ISSUE #4414: Ensure neither the user-defined constructor nor Realm.Object’s constructor is called when returning an accessor object
(i.e. when calling realm.objectForPrimaryKey or iterating a Realm.Result)

### PR DESCRIPTION
**Pull request by kraenhansen**
## What, How & Why?

As a continuation of https://github.com/realm/realm-js/pull/4427 this closes #4414 for our JSI / Hermes implementation.

## ☑️ ToDos
* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [x] 🚦 Tests
* [x] 📱 Check the React Native/other sample apps work if necessary
* [x] 📝 Public documentation PR created or is not necessary
* [x] 💥 `Breaking` label has been applied or is not necessary

### CODE REVIEW

**Review by takameyer — APPROVED** [review]
LGTM

**Comment by kraenhansen** [comment]
I rebased this since the `kh/class-models` has merged into `master` and `hermes` got rebased on those change.

**Review by tomduncalf — APPROVED** [review]
Looks great!

**Review by kneth — APPROVED** [review]
My comments are blocking but mostly clarifying questions

File: integration-tests/tests/src/schemas/person-and-dogs.ts [hunk 1]
```diff
@@ -38,16 +38,13 @@ export const PersonSchema: Realm.ObjectSchema = {
 };
 
 export class Person extends Realm.Object {
```
Comment (takameyer) [inline integration-tests/tests/src/schemas/person-and-dogs.ts:40]: Isn't this supposed to extend `Realm.Object<Person>`?

File: integration-tests/tests/src/tests/class-models.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+          name: "Person",
+          properties: { name: "string" },
+        };
+      }
+      new Realm({ schema: [Person] });
+    });
+  });
+
+  describe("#constructor", () => {
+    type UnmanagedPerson = Partial<Person> & Pick<Person, "name">;
```
Comment (takameyer) [inline integration-tests/tests/src/tests/class-models.ts:78]: Shouldn't we also test the usage of the `Unmanaged` type?

**Review by takameyer — APPROVED** [review]
Just a couple comments

File: integration-tests/tests/src/schemas/person-and-dogs.ts (hunk 1 above)
Comment (kraenhansen) [reply integration-tests/tests/src/schemas/person-and-dogs.ts:40]: If you're not using the constructor, the type argument isn't needed.

File: integration-tests/tests/src/tests/class-models.ts (hunk 2 above)
Comment (kraenhansen) [reply integration-tests/tests/src/tests/class-models.ts:78]: I'll defer that to https://github.com/realm/realm-js/issues/4415.

**Comment by kraenhansen** [comment]
Merging this even with the "Realm Web" integration tests pending: It's are pending download of the BaaS docker image, I've tried re-running it and tests are passing when ran locally.

**Pull request merged**